File: frontend/lib/data/services/notification_service.dart
import '../models/notification.dart';
import '../../core/constants/api_constants.dart';
import 'api_service.dart';

class NotificationService {
  final ApiService _api = ApiService();

  /// Get all notifications for the current user
  Future<List<AppNotification>> getAll() async {
    final response = await _api.get(ApiConstants.notifications);
    return (response as List)
        .map((e) => AppNotification.fromJson(e))
        .toList();
  }

  /// Get unread notifications for the current user
  Future<List<AppNotification>> getUnread() async {
    final response = await _api.get('${ApiConstants.notifications}/unread');
    return (response as List)
        .map((e) => AppNotification.fromJson(e))
        .toList();
  }

  /// Get count of unread notifications
  Future<int> getUnreadCount() async {
    final response = await _api.get('${ApiConstants.notifications}/unread/count');
    return response as int;
  }

  /// Mark a single notification as read
  Future<void> markAsRead(int notificationId) async {
    await _api.patch('${ApiConstants.notifications}/$notificationId/read');
  }

  /// Mark all notifications as read
  Future<void> markAllAsRead() async {
    await _api.patch('${ApiConstants.notifications}/read-all');
  }
}

File: front-angular/src/app/services/notification.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, Observable, interval, switchMap, tap } from 'rxjs';
import { environment } from '../../environments/environment';
import { AppNotification } from '../models/notification.model';

@Injectable({
  providedIn: 'root'
})
export class NotificationService {
  private apiUrl = `${environment.apiUrl}/notifications`;
  private unreadCountSubject = new BehaviorSubject<number>(0);
  public unreadCount$ = this.unreadCountSubject.asObservable();

  constructor(private http: HttpClient) {}

  getAll(): Observable<AppNotification[]> {
    return this.http.get<AppNotification[]>(this.apiUrl);
  }

  getUnread(): Observable<AppNotification[]> {
    return this.http.get<AppNotification[]>(`${this.apiUrl}/unread`);
  }

  getUnreadCount(): Observable<{ count: number }> {
    return this.http.get<{ count: number }>(`${this.apiUrl}/unread/count`).pipe(
      tap(res => this.unreadCountSubject.next(res.count))
    );
  }

  markAsRead(id: number): Observable<AppNotification> {
    return this.http.patch<AppNotification>(`${this.apiUrl}/${id}/read`, null);
  }

  markAllAsRead(): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/read-all`, null).pipe(
      tap(() => this.unreadCountSubject.next(0))
    );
  }

  startPolling(): void {
    interval(15000).pipe(
      switchMap(() => this.getUnreadCount())
    ).subscribe();
  }
}
